command/agent/config.go

```diff
@@ -658,137 +663,116 @@ func (c *Config) Merge(b *Config) *Config {
 	return &result
 }
 
-// normalize config to set defaults and prevent nil derefence panics.
+// normalize Addresses and AdvertiseAddrs to always be initialized and have
+// sane defaults.
+func (c *Config) normalizeAddrs() error {
+	c.Addresses.HTTP = normalizeBind(c.Addresses.HTTP, c.BindAddr)
+	c.Addresses.RPC = normalizeBind(c.Addresses.RPC, c.BindAddr)
+	c.Addresses.Serf = normalizeBind(c.Addresses.Serf, c.BindAddr)
+	c.normalizedAddrs = &Addresses{
+		HTTP: fmt.Sprintf("%s:%d", c.Addresses.HTTP, c.Ports.HTTP),
+		RPC:  fmt.Sprintf("%s:%d", c.Addresses.RPC, c.Ports.RPC),
+		Serf: fmt.Sprintf("%s:%d", c.Addresses.Serf, c.Ports.Serf),
+	}
+
+	addr, err := normalizeAdvertise(c.AdvertiseAddrs.HTTP, c.Addresses.HTTP, c.Ports.HTTP, c.DevMode)
+	if err != nil {
+		return fmt.Errorf("Failed to parse HTTP advertise address: %v", err)
+	}
+	c.AdvertiseAddrs.HTTP = addr
```

Comment:
Fine for this PR but I would have imagined having `normalizedBindAddrs` and `normalizedAdvertiseAddrs` and never modifying the user supplied configs.